plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt

```diff
@@ -0,0 +1,39 @@
+// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+
+package org.jetbrains.kotlin.j2k
+
+import com.intellij.openapi.extensions.ExtensionPointName
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiJavaFile
+
+/**
+ * The `org.jetbrains.kotlin.j2kPreprocessorExtension` extension point enables running custom preprocessing steps on Java files before they
+ * are converted to Kotlin. At runtime, all registered extensions are collected and executed sequentially. To implement your own
+ * preprocessor in a separate plugin, simply extend this interface and register the extension point in your plugin's xml file, e.g.
+ * ```
+ * <extensions defaultExtensionNs="org.jetbrains.kotlin">
+ *   <j2kPreprocessorExtension implementation="org.jetbrains.kotlin.j2k.FooPreprocessorExtension"/>
+ * </extensions>
+ * ```
+ *
+ * All preprocessors are run on a background thread, so write actions must be wrapped in
```

Comment:
Probably worth to mention that they are run in coroutines and therefore must use coroutine-compatible API